### PR TITLE
Fixes issue with goals with losing goals

### PR DESCRIPTION
**Pull request by jannikb**
When executing many actions in a quick succession goals could get incorrectly marked as lost. This causes send-goal-and-wait and wait-for-return to return instantly without waiting for the goal to finish.

To replicate this issue run ```rosrun actionlib_tutorials fibonacci_server``` and in a repl
```lisp
(ros-load:load-system "actionlib_lisp" "actionlib-test")
(in-package :actionlib-lisp)
(start-ros-node "foo")
(defparameter *client* (make-simple-action-client "/fibonacci" "actionlib_tutorials/FibonacciAction"))
(let ((time (get-internal-real-time)))
                  (loop for i from 0 to 100 do
                    (send-goal-and-wait *client* (make-message "actionlib_tutorials/FibonacciGoal" order 0) 2 2)                    
                    (unless (and (eql (state *client*) :SUCCEEDED)
                                 (result *client*))
                      (format t "state: ~a~%" (state *client*))
                      (format t "result: ~a~%" (result *client*))))
                  (- (get-internal-real-time) time))
```

### CODE REVIEW

File: actionlib_lisp/src/new_implementation/comm-state-machine.lisp [hunk 1]
```diff
@@ -146,30 +142,31 @@
...
-  (when (not (eql (latest-goal-status csm) status))
-    (with-recursive-lock ((status-mutex csm))
-      (setf (latest-goal-status csm) status)))
-  (if (get-next-state (stm csm) status)
-        (transition-to csm status)))
+  (with-recursive-lock ((csm-mutex csm))
+    (unless (eql status :lost)
+      (setf (lost-ctr csm) 0))
+    (when (get-next-state (stm csm) status)
+      (setf (latest-goal-status csm) status))
```
Comment (airballking) [inline actionlib_lisp/src/new_implementation/comm-state-machine.lisp]: This method does not compile. I guess the error is in line 156: One closing bracket too much. @jannikb Do I see this correctly?

**Comment by airballking** [comment]
I confirm that the above example code provokes errors with the existing (pre-pull request) implementation.

**Comment by jannikb** [comment]
Should be fixed now.

**Comment by fairlight1337** [comment]
So, can we get this in?
I see serious trouble with using high-frequency TF2 transform lookups:
```lisp
`Action call did not succeed.'
```

**Comment by airballking** [comment]
I am working on it, right now...

**Comment by fairlight1337** [comment]
Just tried it from Jannik's fork, works nicely.

File: actionlib_lisp/src/new_implementation/comm-state-machine.lisp [hunk 2]
```diff
@@ -146,30 +142,31 @@
    transition-callback. If the result was processed before the 
    last status update the transition-callback gets called even
    if the state-machine doesn't change"
-  (if (and (or (eql (name (get-current-state (stm csm))) :done)
-               (process-signal (stm csm) signal))
-           (transition-cb csm))
-      (funcall (transition-cb csm))))
+  (when (and (or (eql (name (get-current-state (stm csm))) :done)
```
Comment (airballking) [inline actionlib_lisp/src/new_implementation/comm-state-machine.lisp]: @jannikb I am repeatedly stumbling over this piece of code. I am just pointing this out as a point where we could improve readability.. I am aware that it was almost like this in the old version.

**Comment by airballking** [comment]
@jannikb I just tried your tests on our robot and it runs into 1~2 lost goals pretty often. Do you also have this?

**Comment by jannikb** [comment]
@airballking In my last tests I didn't have that problem anymore. Is there a way I could reproduce it?

**Comment by airballking** [comment]
I realized I was not testing with the newest version of ```roslisp``` which has the fix for ```ros-time```.

After pulling the amount of lost goals is drastically reduced. Still, I get this:

```lisp
(let ((time (get-internal-real-time)))
  (loop for i from 0 to 5000 do
    (send-goal-and-wait *client* (make-message "actionlib_tutorials/FibonacciGoal" order 0) 2 2)                    
    (unless (and (eql (state *client*) :SUCCEEDED)
                         (result *client*))
      (format t "i=~a~%" i)
      (format t "state: ~a~%" (state *client*))
      (format t "result: ~a~%" (result *client*))))
  (- (get-internal-real-time) time))

[(ACTIONLIB) INFO] 1429266651.608: Reached execute timeout.
i=899
state: LOST
result: [FIBONACCIRESULT
           SEQUENCE:
             #(0 1)]
[(ACTIONLIB) INFO] 1429266664.448: Reached execute timeout.
i=3180
state: LOST
result: [FIBONACCIRESULT
           SEQUENCE:
             #(0 1)]
```

Maybe we can look at this when we meet this afternoon?

**Comment by airballking** [comment]
I started playing with the threshold in the code that declares goals as lost:

https://github.com/jannikb/roslisp_common/blob/master/actionlib_lisp/src/new_implementation/goal-manager.lisp#L143

By increasing this number to 30 I could get rid of the lost goals. Decrementing it to 15, 10, 5, 2 gets me to see a lot of lost goals. 

So, I guess that threshold is a critical parameter. Could we expose it as a member of the action-client?

**Comment by fairlight1337** [comment]
Okay, apparently I've been too optimistic.
```lisp
[(PR2-NAV PROCESS-MODULE) INFO] 1429274644.050: Using nav-pcontroller.
[(PR2-NAV PROCESS-MODULE) INFO] 1429274644.084: Nav action finished.
[(PR2-NAV PROCESS-MODULE) INFO] 1429274644.086: Navigation finished.
[(ACTIONLIB) INFO] 1429274669.974: Reached execute timeout.
[(ACTIONLIB) INFO] 1429274678.390: Reached execute timeout.
```
This happens every time I run navigation (via `(at-location ...)`, specifically). I'd be very interested in the exposure of the parameter @airballking talked about.

**Comment by airballking** [comment]
Note to myself: I started using Jannik's fix for the RoboHow demo on Apr24th. So far, all fine.

**Comment by airballking** [comment]
Just tested this again and it worked for me. I am merging this now in the hope that we ironed out all issues.

**Pull request merged**